### PR TITLE
GH-1634: Define a total, stable sort order.

### LINKED ISSUE #1634: Sorting fails with custom datatype literal
### Version

4.7.0-SNAPSHOT

### What happened?

Dataset: some DBpedia stuff

Query ("give me longest river"):
``` sparql
SELECT DISTINCT ?uri WHERE {
  ?uri a <http://dbpedia.org/ontology/River> 
  {
    ?uri <http://dbpedia.org/ontology/length> ?l 
  } UNION {
    ?uri <http://dbpedia.org/property/length> ?l 
  } 
}
ORDER BY DESC(?l) OFFSET 0 LIMIT 1
```
it fails with 500 and "Comparison method violates its general contract!".

DBpedia data is annoying anyways, there are plenty of length literals with different datatype. For debugging I reduced the literal mix to two :`xsd:double,  	<http://dbpedia.org/datatype/kilometre>`

``` sparql
PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT DISTINCT ?uri WHERE {
  ?uri a <http://dbpedia.org/ontology/River> 
  {
    ?uri <http://dbpedia.org/ontology/length> ?l 
  } UNION {
    ?uri <http://dbpedia.org/property/length> ?l 
  } 
  filter(datatype(?l) in (xsd:double,  	<http://dbpedia.org/datatype/kilometre>))
}
ORDER BY DESC(?l) OFFSET 0 LIMIT 1
```
which still fails.

It does not fail if we project/select the length literal `?l` though:

```sparql
SELECT DISTINCT ?uri ?l WHERE {
  ?uri a <http://dbpedia.org/ontology/River> 
  {
    ?uri <http://dbpedia.org/ontology/length> ?l 
  } UNION {
    ?uri <http://dbpedia.org/property/length> ?l 
  } 
  filter(datatype(?l) in (xsd:double,  	<http://dbpedia.org/datatype/kilometre>))
}
ORDER BY DESC(?l) OFFSET 0 LIMIT 1
```

Looking for the Java exception I found source like here: https://bugs.openjdk.org/browse/JDK-8234482
which also does contain a hint that timsort needs sufficient amount of data to fail:
> TimSort doesn't throw this exception in all cases, though, only if there are a sufficiently large number of elements to be merged (generally > 32 elements, often hundreds are required), and if the sorting algorithm happens to detect an contradiction in the comparison method.

The JDK version is Java 11 by the way if that matters.

### Relevant output and stacktrace

```shell
Query = PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>  SELECT DISTINCT ?uri WHERE {   ?uri a <http://dbpedia.org/ontology/River>    {     ?uri <http://dbpedia.org/ontology/length> ?l    } UNION {     ?uri <http://dbpedia.org/property/length> ?l    }    filter(datatype(?l) in (xsd:double,  	<http://dbpedia.org/datatype/kilometre>)) } ORDER BY DESC(?l) OFFSET 0 LIMIT 1
10:12:42 WARN  Fuseki          :: [32] RC = 500 : Comparison method violates its general contract!
java.lang.IllegalArgumentException: Comparison method violates its general contract!
	at java.util.TimSort.mergeHi(TimSort.java:903) ~[?:?]
	at java.util.TimSort.mergeAt(TimSort.java:520) ~[?:?]
	at java.util.TimSort.mergeCollapse(TimSort.java:448) ~[?:?]
	at java.util.TimSort.sort(TimSort.java:245) ~[?:?]
	at java.util.Arrays.sort(Arrays.java:1441) ~[?:?]
	at org.apache.jena.atlas.data.AbortableComparator.abortableSort(AbortableComparator.java:57) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.atlas.data.SortedDataBag.iterator(SortedDataBag.java:205) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.atlas.data.SortedDataBag.iterator(SortedDataBag.java:192) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.sparql.engine.iterator.QueryIterSort$SortedBindingIterator.initializeIterator(QueryIterSort.java:88) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.atlas.iterator.IteratorDelayedInitialization.init(IteratorDelayedInitialization.java:38) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.atlas.iterator.IteratorDelayedInitialization.hasNext(IteratorDelayedInitialization.java:48) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.sparql.engine.iterator.QueryIterPlainWrapper.hasNextBinding(QueryIterPlainWrapper.java:59) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.sparql.engine.iterator.QueryIteratorBase.hasNext(QueryIteratorBase.java:116) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.sparql.engine.iterator.QueryIterConvert.hasNextBinding(QueryIterConvert.java:58) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.sparql.engine.iterator.QueryIteratorBase.hasNext(QueryIteratorBase.java:116) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.sparql.engine.iterator.QueryIterDistinct.getInputNextUnseen(QueryIterDistinct.java:113) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.sparql.engine.iterator.QueryIterDistinct.hasNextBinding(QueryIterDistinct.java:72) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.sparql.engine.iterator.QueryIteratorBase.hasNext(QueryIteratorBase.java:116) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.sparql.engine.iterator.QueryIterSlice.hasNextBinding(QueryIterSlice.java:76) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.sparql.engine.iterator.QueryIteratorBase.hasNext(QueryIteratorBase.java:116) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.sparql.engine.iterator.QueryIteratorWrapper.hasNextBinding(QueryIteratorWrapper.java:38) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.sparql.engine.iterator.QueryIteratorBase.hasNext(QueryIteratorBase.java:116) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.sparql.engine.iterator.QueryIteratorWrapper.hasNextBinding(QueryIteratorWrapper.java:38) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.sparql.engine.iterator.QueryIteratorBase.hasNext(QueryIteratorBase.java:116) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.sparql.exec.RowSetStream.hasNext(RowSetStream.java:47) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.sparql.engine.ResultSetStream.hasNext(ResultSetStream.java:81) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.fuseki.servlets.SPARQLQueryProcessor.executeQuery(SPARQLQueryProcessor.java:380) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.fuseki.servlets.SPARQLQueryProcessor.execute(SPARQLQueryProcessor.java:279) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.fuseki.servlets.SPARQLQueryProcessor.executeWithParameter(SPARQLQueryProcessor.java:224) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.fuseki.servlets.SPARQLQueryProcessor.execute(SPARQLQueryProcessor.java:209) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.fuseki.servlets.ActionService.executeLifecycle(ActionService.java:58) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.fuseki.servlets.SPARQLQueryProcessor.execPost(SPARQLQueryProcessor.java:84) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.fuseki.servlets.ActionProcessor.process(ActionProcessor.java:34) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.fuseki.servlets.ActionBase.process(ActionBase.java:54) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.fuseki.servlets.ActionExecLib.execActionSub(ActionExecLib.java:124) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.fuseki.servlets.ActionExecLib.execAction(ActionExecLib.java:98) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.fuseki.server.Dispatcher.dispatchAction(Dispatcher.java:164) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.fuseki.server.Dispatcher.process(Dispatcher.java:156) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
	at org.apache.jena.fuseki.server.Dispatcher.dispatch(Dispatcher.java:83) ~[fuseki-server.jar:4.7.0-SNAPSHOT]
```
### Steps to reproduce
* unzip attachment
```bash
sparql --data issue_1634.ttl --query issue_1634.rq
```

[issue_1634.zip](https://github.com/apache/jena/files/10084097/issue_1634.zip)

### Are you interested in making a pull request?

_No response_

### LINKED ISSUE #1692: Improvements to "adjust to timezone"
### Version

4.6.1

### Feature

Add `afn:adjust-to-timezone` which cover all three of `fn:adjust--dateTime-to-timezone`, `fn:adjust--date-to-timezone`, `fn:adjust-time-to-timezone`.

Add some tests.

Fix some formatting in `XSDFuncOp`.

### Are you interested in contributing a solution yourself?

Yes

### PR DESCRIPTION
**Pull request by afs**
GitHub issue resolved #1634
GitHub issue resolved #1692

Pull request Description:
Rework the compare/sorting code.

There is now a complete ordering of based on SPARQL rules, then value then RDF terms.

There will be changes in applications for ORDER BY where previously it was undefined between un-alike terms.
Prior to this set of changes, such sorting was unstable and sometimes threw an exception (#1634).

----

 - [x] Tests are included.
 - [x] Commits have been squashed to remove intermediate development commit messages.
 - [x] Key commit messages start with the issue number (GH-xxxx or JENA-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

### CODE REVIEW

File: jena-arq/src/main/java/org/apache/jena/sparql/expr/NodeValueCmp.java [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+
+    /**
+     * The sort order is to apply the rules in the following order:
+     * <ol>
+     * <li>Order by value space
+     * <li>Within a value space, order by value (if possible)
+     * <li>Order by RDF term (syntax)
+     * </ol>
+     * <p>
+     * The implicit timezone for dateTime/date etc is fixed UTC to ensure sortingis teh same everywhere.
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/sparql/expr/NodeValueCmp.java:399]:
```suggestion
     * The implicit timezone for dateTime/date etc is fixed UTC to ensure sorting is the same everywhere.
```

File: jena-arq/src/main/java/org/apache/jena/sparql/expr/nodevalue/XSDFuncOp.java [hunk 2]
```diff
@@ -1131,118 +1131,74 @@ public static int compareString(NodeValue nv1, NodeValue nv2) {
     // works for dates as well because they are implemented as dateTimes on their start point.
 
     public static int compareDateTime(NodeValue nv1, NodeValue nv2) {
+        // XSD Datatypes defines partial order on dateTies when comparing with and
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/sparql/expr/nodevalue/XSDFuncOp.java:1134]:
```suggestion
        // XSD Datatypes defines partial order on dateTimes when comparing with and
```

File: jena-arq/src/main/java/org/apache/jena/sparql/expr/nodevalue/XSDFuncOp.java [hunk 3]
```diff
@@ -1131,118 +1131,74 @@ public static int compareString(NodeValue nv1, NodeValue nv2) {
...
+        // with timezone => compareXSDDateTime
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/sparql/expr/nodevalue/XSDFuncOp.java:1135]:
```suggestion
        // without timezone => compareXSDDateTime
```

File: jena-arq/src/test/java/org/apache/jena/sparql/expr/TestComparison.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    private static void compare(String string1, String string2, int expected) {
+        compare(nodeValue(string1), nodeValue(string2), expected);
+    }
+
+    private static void compare(NodeValue nv1, NodeValue nv2, int expected) {
+        try {
+            //if ( true ) throw new RuntimeException("Switch to sorting branch");
+            //int cmp2 = -99;
+            int cmp2 = NodeValueCmp.compareByValue(nv1, nv2);
+
```
Comment (rvesse) [inline jena-arq/src/test/java/org/apache/jena/sparql/expr/TestComparison.java:68]: No assertion that the comparison result is as expected?

**Review by rvesse — COMMENTED** [review]
Mostly looks good, couple of minor typos and one set of tests that appears to be missing actual assertions

On the whole I think it is better to move to having a fully stable sort ordering in place so users aren't getting seemingly random failures (and sometimes subtly wrong ordering) even if it potentially changes some existing result orderings

File: jena-arq/src/test/java/org/apache/jena/sparql/expr/TestComparison.java (hunk 4 above)
Comment (afs) [reply jena-arq/src/test/java/org/apache/jena/sparql/expr/TestComparison.java]: Now checks it is the expected value.

All other suggestions done..

**Pull request merged**